### PR TITLE
Arsenal - Add Default Loadouts API

### PR DESCRIPTION
**Pull request by 654wak654**
**When merged this pull request will:**
- Add a `addDefaultLoadout` function to arsenal so addon and mission makers can add default loadouts through script easier.
- Change other loadout related arsenal functions to use `findIf` like the new function.

### CODE REVIEW

File: addons/arsenal/XEH_preInit.sqf [hunk 1]
```diff
@@ -8,6 +8,7 @@ PREP_RECOMPILE_START;
 PREP_RECOMPILE_END;
 
 // Arsenal
+GVAR(defaultLoadoutsList) = [];
```
Comment (alganthe) [inline addons/arsenal/XEH_preInit.sqf]: You're breaking default loadouts, this variable already exists whatever happens due to https://github.com/acemod/ACE3/blob/master/addons/arsenal/Cfg3DEN.hpp#L11

File: addons/arsenal/functions/fnc_addDeafultLoadout.sqf [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ *
+ * Example:
+ * ["Squad Leader", getUnitLoadout sql1] call ace_arsenal_fnc_addDefaultLoadout
+ *
+ * Public: Yes
+*/
+
+params [["_name", "", [""]], ["_loadout", [], [[]], 10]];
+
+GVAR(defaultLoadoutsList) append [_name, _loadout];
```
Comment (dedmen) [inline addons/arsenal/functions/fnc_addDeafultLoadout.sqf]: Maybe some check to not add same name twice? Not sure if that matters.

File: addons/arsenal/functions/fnc_buttonExport.sqf [hunk 3]
```diff
@@ -16,7 +16,7 @@ params ["_display"];
 
 if (GVAR(shiftState)) then {
 
-    if (isNil QGVAR(defaultLoadoutsList) || {GVAR(defaultLoadoutsList) isEqualTo []}) exitWith {
```
Comment (dedmen) [inline addons/arsenal/functions/fnc_buttonExport.sqf]: This is unscheduled anyway so nil shouldn't matter right?

File: addons/arsenal/functions/fnc_addDeafultLoadout.sqf (hunk 2 above)
Comment (alganthe) [reply addons/arsenal/functions/fnc_addDeafultLoadout.sqf]: true, completely forgot about that too, you definitely need name checking in there.

File: addons/arsenal/functions/fnc_onArsenalOpen.sqf [hunk 4]
```diff
@@ -34,12 +34,8 @@ if (isNil {GVAR(sharedLoadoutsNamespace) getVariable QGVAR(sharedLoadoutsVars)})
...
 };
 
-if (isNil QGVAR(defaultLoadoutsList)) then {
-    if (is3DEN) then {
-        GVAR(defaultLoadoutsList) = (QGVAR(DummyCategory) get3DENMissionAttribute QGVAR(DefaultLoadoutsListAttribute));
-    } else {
-            GVAR(defaultLoadoutsList) = [];
-    };
+if (is3DEN) then {
+    GVAR(defaultLoadoutsList) append (QGVAR(DummyCategory) get3DENMissionAttribute QGVAR(DefaultLoadoutsListAttribute));
```
Comment (dedmen) [inline addons/arsenal/functions/fnc_onArsenalOpen.sqf]: parenthesis for readability.

File: addons/arsenal/functions/fnc_buttonExport.sqf (hunk 3 above)
Comment (alganthe) [reply addons/arsenal/functions/fnc_buttonExport.sqf]: it can't be nil, so no worries. (my mistake)

**Comment by dedmen** [comment]
<https://github.com/acemod/ACE3/pull/6558#discussion_r214539698>
No it's not. preInit runs before the 3DEN code. Meaning either the 3DEN code doesn't run and the list is empty. or it runs and overwrites the old value.

File: addons/arsenal/functions/fnc_addDeafultLoadout.sqf (hunk 2 above)
Comment (alganthe) [reply addons/arsenal/functions/fnc_addDeafultLoadout.sqf]: I can't identify loadouts by magic, I had to use something (plus it makes sense)

**Comment by bux** [comment]
lacks documentation :)

File: addons/arsenal/Cfg3DEN.hpp [hunk 5]
```diff
@@ -8,7 +8,7 @@ class Cfg3DEN {
                         class GVAR(DefaultLoadoutsListAttribute) {
                             property = QGVAR(DefaultLoadoutsListAttribute);
                             value = 0;
-                            expression = "if !(is3DEN) then {ace_arsenal_defaultLoadoutsList = _value};";
```
Comment (654wak654) [inline addons/arsenal/Cfg3DEN.hpp]: @alganthe Was there a reason this was like this? Could I break something by using QUOTE?

Comment (commy2) [reply addons/arsenal/Cfg3DEN.hpp]:
```cpp
expression = QUOTE(if !(is3DEN) then {GVAR(defaultLoadoutsList) = _value});
```
works, and so does
```cpp
expression = if !(is3DEN) then {GVAR(defaultLoadoutsList) = _value};
```

Comment (654wak654) [reply addons/arsenal/Cfg3DEN.hpp]: I'll keep the verbose one with `QUOTE`, but good point. Might abuse it later.

Comment (jonpas) [reply addons/arsenal/Cfg3DEN.hpp]: Second one just gets auto-quoted, and will throw missing quotations warning with armake (unless that warning is disabled).

File: addons/arsenal/functions/fnc_buttonLoadoutsSave.sqf [hunk 6]
```diff
@@ -155,7 +155,7 @@ switch (GVAR(currentLoadoutsTab)) do {
 
         if (is3DEN) then {
 
-            private _sameNameLoadoutsList = _data select {_x select 0 == _editBoxContent};
+            private _loadoutIndex = (+(_data)) findId {(_x select 0) == _editBoxContent};
```
Comment (Dystopian) [inline addons/arsenal/functions/fnc_buttonLoadoutsSave.sqf]: `findId` -> `findIf` ?

Comment (654wak654) [reply addons/arsenal/functions/fnc_buttonLoadoutsSave.sqf]: RIP

Comment (dedmen) [reply addons/arsenal/functions/fnc_buttonLoadoutsSave.sqf]: Why `(+(_data))` ? Why copy the array. Don't see any need.

Comment (654wak654) [reply addons/arsenal/functions/fnc_buttonLoadoutsSave.sqf]: @alganthe had me deep copy everything

Comment (dedmen) [reply addons/arsenal/functions/fnc_buttonLoadoutsSave.sqf]: *slaps @alganthe*

Comment (alganthe) [reply addons/arsenal/functions/fnc_buttonLoadoutsSave.sqf]: hey no, I didn't told you to deep copy EVERYTHING, just the namespace var that is used to store loadouts (depends on which tab is active).

**Comment by dedmen** [comment]
Just tried to use this and wondered why it doesn't work. Turns out, it's not released yet :U I need this tho

**Pull request merged**